frontend/src/Components/FavoriteCard.tsx

```diff
@@ -0,0 +1,122 @@
+import { useNavigate } from 'react-router-dom';
+import { StarIcon as SolidStar } from '@heroicons/react/24/solid';
+import { mutate } from 'swr';
+import { ToastState, OpenContentFavorite, ServerResponseMany } from '@/common';
+import API from '@/api/api';
+import { useToast } from '@/Context/ToastCtx';
+
+interface FavoriteCardProps {
+    favorite: OpenContentFavorite;
+    pageQuery: number;
+    perPage: number;
+}
+
+const FavoriteCard: React.FC<FavoriteCardProps> = ({ favorite }) => {
+    const navigate = useNavigate();
+    const { toaster } = useToast();
+
+    const handleCardClick = () => {
+        if (favorite.visibility_status) {
+            return;
+        }
+        const state = {
+            subLink: favorite.content_url
+        };
+        if (favorite.type === 'video') {
+            navigate(`/viewer/videos/${favorite.content_id}`);
+        } else if (favorite.type === 'library') {
+            navigate(`/viewer/libraries/${favorite.content_id}`, { state });
+        }
+    };
+
+    const handleUnfavorite = async () => {
+        const endpoint =
+            favorite.type === 'video'
+                ? `videos/${favorite.content_id}/favorite`
+                : `open-content/${favorite.content_id}/save`;
+
+        const { name, open_content_provider_id } = favorite;
+
+        if (!name || !open_content_provider_id) {
+            throw new Error('Favorite data is incomplete.');
+        }
+
+        const payload = {
+            name,
+            content_id: favorite.content_id,
+            open_content_provider_id,
+            content_url: `/api/proxy/libraries/${favorite.content_id}/`
+        };
+
+        const response = await API.put(endpoint, payload);
+        if (response.success) {
+            toaster(`${name} removed from favorites`, ToastState.success);
+
+            await mutate(
+                (key: string | null | undefined): boolean => {
+                    if (typeof key === 'string') {
+                        return key.includes('open-content/favorites');
+                    }
+                    return false;
+                },
+                (
+                    cachedData:
+                        | ServerResponseMany<OpenContentFavorite>
+                        | undefined
+                ) => {
+                    if (!cachedData) return;
+
+                    const newFavorites = cachedData.data.filter(
+                        (item: OpenContentFavorite) =>
+                            item.content_id !== favorite.content_id
+                    );
+
+                    return {
+                        ...cachedData,
+                        data: newFavorites
+                    };
+                },
+                false
+            );
+
+            if (favorite.type === 'library') {
+                await mutate('/api/libraries');
+            }
```

Comment:
there are two ways to do this which would condense this code a bunch:
1. I listed this in my previous comment, but you could have the `mutate` attached to the initial useSWR call that calls the favorites api and pass that function into this favorite card, and then simply just call `await mutate()` on success. Because it is established on the original useSWR call, it knows automatically which api call it has to revalidate.
2. You could skip passing in the mutate into this favorites card and instead you could import the global mutate from SWR and then give the function the api that you want to revalidate. In this case, you would also have to pass in page and per page, which would need to be passed into the card (since i guess it is required in the backend? i haven't checked the backend out, but it seems like it will not return anything if those variables are not defined. although they probably should by default be something i think...). So, the call would be one simple line, ``await mutate(`/api/open-content/favorites?page=${pageQuery}&per_page=${perPage}`)``, which will revalidate the data on this favorites page.

The reason you do not need to call mutate on `api/libraries` or videos is because those pages will be revalidated when you switch to those tabs anyways, since they are separate useSWR calls. So, when the user changes from favorites -> libraries tab, that component, while rendering, will take the useSWR hook on that page and call the `api/libraries` handler and will fetch the new, updated data.

Reply:
Also if im completely off base on this and not understanding why you have to do this and it IS a requirement, please let me know!